src/utilities/util_dump.c

```diff
@@ -663,67 +674,123 @@ dump_record(WT_CURSOR *cursor, const char *key, bool reverse, bool search_near,
         suffix = "\n";
     }
 
-    while (ret == 0) {
-
-        /* A specific key is requested. */
-        if (key != NULL) {
-            current_key = key;
-            cursor->set_key(cursor, current_key);
-
-            ret = cursor->search_near(cursor, &exact);
-
-            if (ret != 0 && ret != WT_NOTFOUND)
-                return (util_cerr(cursor, "search_near", ret));
-
-            /*
-             * If a key has been found but search near is disabled and there is no exact match, the
-             * requested key is missing.
-             */
-            if (ret == 0 && !search_near && exact != 0)
-                ret = WT_NOTFOUND;
-
-            /* A key has been found. */
-            if (ret == 0) {
-                if (search_near && exact != 0) {
-                    /* Retrieve the key found by search_near. */
-                    if ((ret = cursor->get_key(cursor, &current_key)) != 0)
-                        return (util_cerr(cursor, "get_key", ret));
-                }
-                if ((ret = cursor->get_value(cursor, &value)) != 0)
-                    return (util_cerr(cursor, "get_value", ret));
-            }
+    if ((ret = cursor->get_key(cursor, &current_key)) != 0)
+        return (util_cerr(cursor, "get_key", ret));
+    if ((ret = cursor->get_value(cursor, &value)) != 0)
+        return (util_cerr(cursor, "get_value", ret));
+    if (fprintf(fp, "%s%s%s%s%s", prefix, current_key, infix, value, suffix) < 0)
+        return (util_err(session, EIO, NULL));
+    return (0);
+}
 
-        } else {
-            /* Continue parsing the file. */
-            ret = reverse ? cursor->prev(cursor) : cursor->next(cursor);
-            if (ret == 0) {
-                if ((ret = cursor->get_key(cursor, &current_key)) != 0)
-                    return (util_cerr(cursor, "get_key", ret));
-                if ((ret = cursor->get_value(cursor, &value)) != 0)
-                    return (util_cerr(cursor, "get_value", ret));
-            } else if (ret != WT_NOTFOUND)
-                return (util_cerr(cursor, (reverse ? "prev" : "next"), ret));
+/*
+ * dump_record --
+ *     Dump the record specified by key or one near to it. If a window is specified print out up to
+ *     that many records before and after sought record. The window will be truncated if it would
+ *     move past the first or last entry.
+ */
+static int
+dump_record(
+  WT_CURSOR *cursor, const char *key, bool reverse, bool search_near, bool json, uint64_t window)
+{
+    WT_DECL_RET;
+    WT_SESSION *session;
+    int exact;
+    const char *current_key;
+    bool once;
+    unsigned int n, total_window;
+    int (*fwd)(WT_CURSOR *);
+    int (*bck)(WT_CURSOR *);
+
+    session = cursor->session;
+    once = false;
+    exact = 0;
+
+    WT_ASSERT((WT_SESSION_IMPL *)session, key != NULL);
+
+    current_key = key;
+    cursor->set_key(cursor, current_key);
+    ret = cursor->search_near(cursor, &exact);
```

Comment:
My preference would be to break the walking into its own function with a function pointer passed to determine direction. We could save the key we found to restart the walk from that point.

Reply:
If the printing is done in a separate function is there any issue with just passing the currently positioned cursor down to the printing function?

Reply:
Due to time constraints I would like to defer any such refactoring.